feat: add Windows support in vsphere-supervisor

dilyar85 opened the pull request:
This PR introduces new `bootstrap_provider` and `bootstrap_data_file` configs to enable deploying Windows VMs in the `vsphere-supervisor` builder. If these configurations are not set, the builder will default to using minimal cloud-config via CloudInit as it currently does.

Deploying a Windows image with Sysprep using the example unattend data file:

```console
$ packer build windows-template.pkr.hcl
==> vsphere-supervisor.vm: Creating temporary RSA SSH key for instance...
    vsphere-supervisor.vm: Connecting to Supervisor cluster...
    vsphere-supervisor.vm: Successfully connected to Supervisor cluster
    vsphere-supervisor.vm: Creating required source objects in Supervisor cluster...
    vsphere-supervisor.vm: Creating a K8s Secret object for providing source VM bootstrap data...
    vsphere-supervisor.vm: Loading bootstrap data from file: /Users/sdiliyaer/packer-plugin-vsphere/builder/vsphere/examples/supervisor/sysprep-unattend-data.yml
    vsphere-supervisor.vm: Successfully created the K8s Secret object
    vsphere-supervisor.vm: Creating a source VirtualMachine object
    vsphere-supervisor.vm: Successfully created the VirtualMachine object
    vsphere-supervisor.vm: Creating a VirtualMachineService object for network connection
    vsphere-supervisor.vm: Successfully created the VirtualMachineService object
    vsphere-supervisor.vm: Finished creating all required source objects in Supervisor cluster
    vsphere-supervisor.vm: Waiting for the source VM to be powered-on and accessible...
    vsphere-supervisor.vm: Source VM is NOT powered-on yet, continue watching...
    vsphere-supervisor.vm: Source VM is powered-on, waiting for an IP to be assigned...
    vsphere-supervisor.vm: Successfully obtained the source VM IP: 192.168.128.40
    vsphere-supervisor.vm: Getting source VM ingress IP from the VMService object
    vsphere-supervisor.vm: Successfully retrieved the source VM ingress IP: 192.168.0.19
    vsphere-supervisor.vm: Source VM is now ready in Supervisor cluster
==> vsphere-supervisor.vm: Using SSH communicator to connect: 192.168.0.19
==> vsphere-supervisor.vm: Waiting for SSH to become available...
==> vsphere-supervisor.vm: Connected to SSH!
==> vsphere-supervisor.vm: Provisioning with Powershell...
==> vsphere-supervisor.vm: Provisioning with powershell script: /var/folders/nf/cndlm4ts2bd9jz3dd667jfb80000gr/T/powershell-provisioner3594983132
    vsphere-supervisor.vm: Skip cleaning up the source objects as specified in config
    vsphere-supervisor.vm: Build 'vsphere-supervisor' finished without publishing the VM image (feature not available yet).
Build 'vsphere-supervisor.vm' finished after 20 minutes 17 seconds.
```

Deploying a Linux image with default configs:

```console
$ packer build ~/packer-templates/linux.pkr.hcl
==> vsphere-supervisor.vm: Creating temporary RSA SSH key for instance...
    vsphere-supervisor.vm: Connecting to Supervisor cluster...
    vsphere-supervisor.vm: Successfully connected to Supervisor cluster
    vsphere-supervisor.vm: Creating required source objects in Supervisor cluster...
    vsphere-supervisor.vm: Creating a K8s Secret object for providing source VM bootstrap data...
    vsphere-supervisor.vm: Using default cloud-init user data as the 'bootstrap_data_file' is not specified
    vsphere-supervisor.vm: Successfully created the K8s Secret object
    vsphere-supervisor.vm: Creating a source VirtualMachine object
    vsphere-supervisor.vm: Successfully created the VirtualMachine object
    vsphere-supervisor.vm: Creating a VirtualMachineService object for network connection
    vsphere-supervisor.vm: Successfully created the VirtualMachineService object
    vsphere-supervisor.vm: Finished creating all required source objects in Supervisor cluster
    vsphere-supervisor.vm: Waiting for the source VM to be powered-on and accessible...
    vsphere-supervisor.vm: Source VM is NOT powered-on yet, continue watching...
    vsphere-supervisor.vm: Source VM is powered-on, waiting for an IP to be assigned...
    vsphere-supervisor.vm: Successfully obtained the source VM IP: 192.168.128.43
    vsphere-supervisor.vm: Getting source VM ingress IP from the VMService object
    vsphere-supervisor.vm: Successfully retrieved the source VM ingress IP: 192.168.0.20
    vsphere-supervisor.vm: Source VM is now ready in Supervisor cluster
==> vsphere-supervisor.vm: Using SSH communicator to connect: 192.168.0.20
==> vsphere-supervisor.vm: Waiting for SSH to become available...
==> vsphere-supervisor.vm: Connected to SSH!
==> vsphere-supervisor.vm: Provisioning with shell script: /var/folders/nf/cndlm4ts2bd9jz3dd667jfb80000gr/T/packer-shell2147188375
    vsphere-supervisor.vm: Skip cleaning up the source objects as specified in config
    vsphere-supervisor.vm: Build 'vsphere-supervisor' finished without publishing the VM image (feature not available yet).
Build 'vsphere-supervisor.vm' finished after 5 minutes 12 seconds.
```

Deploying with `communicator=none` and no provisioners defined in the template:

```console
$ packer build -var communicator=none ~/tmp/test-template.pkr.hcl
vsphere-supervisor.vm: output will be in this color.

    vsphere-supervisor.vm: Connecting to Supervisor cluster...
    vsphere-supervisor.vm: Successfully connected to Supervisor cluster
    vsphere-supervisor.vm: Creating required source objects in Supervisor cluster...
    vsphere-supervisor.vm: Creating a K8s Secret object for providing source VM bootstrap data...
    vsphere-supervisor.vm: Using default cloud-init user data as the 'bootstrap_data_file' is not specified
    vsphere-supervisor.vm: Successfully created the K8s Secret object
    vsphere-supervisor.vm: Creating a source VirtualMachine object
    vsphere-supervisor.vm: Successfully created the VirtualMachine object
    vsphere-supervisor.vm: Skip creating VirtualMachineService as communicator type is 'none'
    vsphere-supervisor.vm: Finished creating all required source objects in Supervisor cluster
    vsphere-supervisor.vm: Waiting for the source VM to be powered-on and accessible...
    vsphere-supervisor.vm: Source VM is NOT powered-on yet, continue watching...
    vsphere-supervisor.vm: Source VM is powered-on, waiting for an IP to be assigned...
    vsphere-supervisor.vm: Successfully obtained the source VM IP: 192.168.128.54
    vsphere-supervisor.vm: Source VM is now ready in Supervisor cluster
    vsphere-supervisor.vm: Skip cleaning up the source objects as specified in config
    vsphere-supervisor.vm: Build 'vsphere-supervisor' finished without publishing the VM image (feature not available yet).
Build 'vsphere-supervisor.vm' finished after 3 minutes 9 seconds.
```

Closes #274

dilyar85 commented:
Hey @nywilken, I tested this change on our internal testbed and added the successful plugin run output to the PR description. Would you mind checking it out when you have a moment? Thanks!

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -112,6 +116,8 @@ func (s *StepWatchSource) initStep(state multistep.StateBag) error {
 	s.SourceName = sourceName
 	s.Namespace = namespace
 	s.KubeWatchClient = kubeWatchClient
+	s.timeoutCh = time.After(time.Duration(s.Config.WatchSourceTimeoutSec) * time.Second)
```
I believe that storing this during the initStep actually starts the timer, which mean that reusing it won't actually give you the full duration. Have you tested this with a smaller duration to see how it behaves?

nywilken replied:
We can replace this by using a context that is created within the Run function. I found this to be a bit confusing at first but now understand its usages given our internal chat. 

```suggestion
```

FYI the timer will start upon initialization.

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -165,7 +171,7 @@ func (s *StepWatchSource) waitForVMReady(ctx context.Context, logger *PackerLogg
 			}
 
-		case <-time.After(time.Duration(s.Config.WatchSourceTimeoutSec) * time.Second):
+		case <-s.timeoutCh:
```
To ensure each step gets the full elapsed time. 
```suggestion
		case <-time.After(time.Duration(s.Config.WatchSourceTimeoutSec) * time.Second):
```

dilyar85 replied:
Resolving this comment as it's intended.

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -175,21 +181,33 @@ func (s *StepWatchSource) getVMIngressIP(ctx context.Context, logger *PackerLogg
+			logger.Info("Successfully retrieved the source VM ingress IP: %s", ingress[0].IP)
+			return ingress[0].IP, nil
+
+		case <-s.timeoutCh:
```
To ensure each step gets the full elapsed time. 
```suggestion
		case <-time.After(time.Duration(s.Config.WatchSourceTimeoutSec) * time.Second):
```

dilyar85 replied:
Ditto.

nywilken reviewed (commented):
@dilyar85 I left a few suggestions so far. I have an open question about the use of the timeoutCh. Specifically does the timer start upon initialization or when you start listening to the channel. If it is the former then I think you may be getting a smaller time interval for the second timeout and a potential blocking call if the first timer returns and we hit the second timer. I will review more tomorrow.

dilyar85 commented:
@nywilken Thanks for the review! I've addressed all your comments in a separate commit, hoping it's easier to see the changes. When the PR is ready to merge, I'll squash all my commits into one.

zyiyi11 commented on builder/vsphere/supervisor/step_create_source.go:
```diff
@@ -39,14 +44,18 @@ type CreateSourceConfig struct {
 	NetworkName string `mapstructure:"network_name"`
 	// Preserve the created objects even after importing them to the vSphere endpoint. Defaults to `false`.
 	KeepInputArtifact bool `mapstructure:"keep_input_artifact"`
+	// Name of the bootstrap provider to use for configuring the source VM. Defaults to `CloudInit`.
```
It would helpful to list all supported bootstrap here.

dilyar85 replied:
Updated in the latest commit.

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -49,6 +49,7 @@ type StepWatchSource struct {
 
 	SourceName, Namespace string
 	KubeWatchClient       client.WithWatch
+	timeoutCh             <-chan time.Time
```
I took a little time thinking about this give the provided info on Slack and in this PR, and think we can remove this entirely and change up the code a bit. I'll comment further down in the code with the proposed changes.


```suggestion
```

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -165,7 +171,7 @@ func (s *StepWatchSource) waitForVMReady(ctx context.Context, logger *PackerLogg
 			}
 
-		case <-time.After(time.Duration(s.Config.WatchSourceTimeoutSec) * time.Second):
+		case <-s.timeoutCh:
```
So reading up a little on the use of time.After I found that this is actually not a preferred because the timer channel created by time.After is never closed if it never times out. So on a successful run here the function returns but the timer channel is not closed, which can lead to a memory leak. 

Given your usage the memory leak would not occur here but in the select statement within the getVMIngressIP function. If step.Run succeeds without issue that the timeoutCh is never closed. Which you can fix by using a time.NewTimer and classing timeoutCh.Stop() in the Run function. But I am proposing a slightly different solution using context.WithTimeout

```suggestion
		case <-ctx.Done():
```

nywilken commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -175,21 +181,33 @@ func (s *StepWatchSource) getVMIngressIP(ctx context.Context, logger *PackerLogg
+			logger.Info("Successfully retrieved the source VM ingress IP: %s", ingress[0].IP)
+			return ingress[0].IP, nil
+
+		case <-s.timeoutCh:
```
```suggestion
		case <-ctx.Done():
```

nywilken reviewed (commented):
Hi @dilyar85 thanks again for pushing this forward and for allowing me the time to review/response. I have a few suggestions on the timeout logic given the new information you provided. 

I tired to capture my comments within the review but it is a lot so I created a branch with the suggested changes to help illustrate the change. If you could pull it down to test against your environment that would be a nice way to validate that I did not break anything. 

https://github.com/hashicorp/packer-plugin-vsphere/commit/0d04f05042a3cf1a0a7cbbe5b932e6732e4efa2b

nywilken reviewed (approved):
@dilyar85 thanks for testing and applying the requested changes. This looks good to me. I'll merge and release as 1.2.0 by EOD tomorrow, if not later today.

dilyar85 commented:
> @dilyar85 thanks for testing and applying the requested changes. This looks good to me. I'll merge and release as 1.2.0 by EOD tomorrow, if not later today.

Thanks for merging this PR. In terms of the release, it would be nice if we could include the other supervisor builder change (https://github.com/hashicorp/packer-plugin-vsphere/pull/276) as well. The dependent repo is now public, and I'll work on updating that PR to get it ready shortly.
